dashboard/app/asset_storage.go

```diff
@@ -0,0 +1,347 @@
+		bugKey := key.Parent()
+		if _, ok := ad.relevantBugs[bugKey.String()]; ok {
+			// At least one crash is related to an opened/recently closed bug.
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (ad *assetDeprecator) needThisBuildAsset(build *Build, buildAsset *Asset) (bool, error) {
+	if buildAsset.Type == asset.HTMLCoverageReport {
+		// We want to keep coverage reports forever, not just
+		// while there are any open bugs. But we don't want to
+		// keep all coverage reports, just a share of them.
+		return ad.buildArchivePolicy(build, buildAsset)
+	}
+	if build.Type == BuildNormal {
+		// A build-related asset, keep it only while there are open bugs with crashes
+		// related to this build.
+		return ad.buildBugStatusPolicy(build)
+	}
+	// TODO: fix this once this is no longer the case.
+	return false, fmt.Errorf("job-related assets are not supported yet")
+}
+
+func (ad *assetDeprecator) updateBuild(buildID string, urlsToDelete []string) error {
+	toDelete := map[string]bool{}
+	for _, url := range urlsToDelete {
+		toDelete[url] = true
+	}
+	tx := func(c context.Context) error {
+		build, err := loadBuild(ad.c, ad.ns, buildID)
+		if build == nil || err != nil {
+			// Assume the DB has been updated in the meanwhile.
+			return nil
+		}
+		newAssets := []Asset{}
+		for _, asset := range build.Assets {
+			if _, ok := toDelete[asset.DownloadURL]; !ok {
+				newAssets = append(newAssets, asset)
+			}
+		}
+		build.Assets = newAssets
+		build.AssetsLastCheck = timeNow(ad.c)
+		if _, err := db.Put(ad.c, buildKey(ad.c, ad.ns, buildID), build); err != nil {
+			return fmt.Errorf("failed to save build: %w", err)
+		}
+		return nil
+	}
+	if err := db.RunInTransaction(ad.c, tx, nil); err != nil {
+		return fmt.Errorf("failed to update build: %w", err)
+	}
+	return nil
+}
+
+func (ad *assetDeprecator) batchProcessBuilds(count int) error {
+	// We cannot query only the Build with non-empty Assets array and yet sort
+	// by AssetsLastCheck. The datastore returns "The first sort property must
+	// be the same as the property to which the inequality filter is applied.
+	// In your query the first sort property is AssetsLastCheck but the inequality
+	// filter is on Assets.DownloadURL.
+	// So we have to omit Filter("Assets.DownloadURL>", ""). here.
+	var builds []*Build
+	_, err := db.NewQuery("Build").
+		Filter("Namespace=", ad.ns).
+		Order("AssetsLastCheck").
+		Limit(count).
+		GetAll(ad.c, &builds)
+	if err != nil {
+		return fmt.Errorf("failed to fetch builds: %w", err)
+	}
+	for _, build := range builds {
+		toDelete := []string{}
+		for _, asset := range build.Assets {
+			needed, err := ad.needThisBuildAsset(build, &asset)
+			if err != nil {
+				return fmt.Errorf("failed to test asset: %w", err)
+			} else if !needed {
+				toDelete = append(toDelete, asset.DownloadURL)
+			}
+		}
+		err := ad.updateBuild(build.ID, toDelete)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+const buildBatchSize = 16
```

Comment:
move inside of the function, just before the use
it's used once and the name is generic